client/src/main/scala/org/http4s/client/middleware/Retry.scala

```diff
@@ -0,0 +1,50 @@
+package org.http4s
+package client
+package middleware
+
+import org.http4s.Status.ResponseClass.Successful
+import org.http4s.{Response, Request, EmptyBody}
+import org.http4s.client.Client
+import scala.concurrent.duration._
+import scala.math.{pow, min, random}
+import scalaz.concurrent.Task
+import scala.language.postfixOps
+
+object Retry {
+
+  def apply(backoff: Task[Int => Option[FiniteDuration]])(client: Client) = new Client {
```

Comment:
I originally had an implementation for backoff that was just `Int => Option[FiniteDuration]`. @rossabaker gave me some feedback that putting this inside a Task might be cleaner and would allow to just re-evaluate the Task to get the next attempt.